lsp-mode.el

```diff
@@ -1968,7 +1983,96 @@ WORKSPACE is the workspace that contains the progress token."
+
+(defun lsp--headerline-build-string (symbols-hierarchy)
+  "Build the header-line from SYMBOLS-HIERARCHY."
+  (seq-reduce (lambda (last-symbol-name symbol-to-append)
+                (let ((symbol2-name (if (lsp:document-symbol-deprecated? symbol-to-append)
+                                        (propertize (lsp:document-symbol-name symbol-to-append)
+                                                    'font-lock-face 'lsp-headerline-breadcrumb-deprecated-face)
+                                      (propertize (lsp:document-symbol-name symbol-to-append)
+                                                  'font-lock-face lsp-headerline-breadcrumb-face)))
+                      (symbol2-icon (lsp--headerline-breadcrumb-symbol-icon symbol-to-append))
+                      (arrow-icon (lsp--headerline-breadcrumb-arrow-icon)))
+                  (format "%s %s %s"
+                          last-symbol-name
+                          arrow-icon
+                          (if symbol2-icon
+                              (concat symbol2-icon symbol2-name)
+                            symbol2-name))))
+              symbols-hierarchy ""))
+
+(defun lsp--headerline-document-symbols->symbols-hierarchy (document-symbols)
+  "Convert DOCUMENT-SYMBOLS to symbols hierarchy."
+  (-let (((symbol &as &DocumentSymbol? :children?) (seq-some (-lambda ((symbol &as &DocumentSymbol :range))
+                                                          (-let (((beg . end) (lsp--range-to-region range)))
+                                                            (and (<= beg (point) end)
+                                                                 symbol)))
+                                                        document-symbols)))
+    (if children?
+        (append (list symbol) (lsp--headerline-document-symbols->symbols-hierarchy children?))
+      (when symbol
+        (list symbol)))))
+
+(defun lsp--headerline-symbols-informations->symbols-hierarchy (symbols-informations)
+  "Convert SYMBOL-INFORMATIONS to symbols hierarchy."
+  (seq-some (-lambda ((symbol &as &SymbolInformation :location (&Location :range)))
+              (-let (((beg . end) (lsp--range-to-region range)))
+                (and (<= beg (point) end)
+                     symbol)))
+            symbols-informations))
+
+(defun lsp--headerline-symbols->symbols-hierarchy (symbols)
+  "Convert SYMBOLS to symbols-hierarchy."
+  (when-let (first-symbol (lsp-seq-first symbols))
+    (if (lsp-symbol-information? first-symbol)
+        (lsp--headerline-symbols-informations->symbols-hierarchy symbols)
+      (lsp--headerline-document-symbols->symbols-hierarchy symbols))))
+
+(defun lsp--headerline-check-breadcrumb (&rest _)
+  "Request for document symbols to build the breadcrumb."
+  (when (lsp-feature? "textDocument/documentSymbol")
+    (-if-let* ((lsp--document-symbols-request-async t)
+               (symbols (lsp--get-document-symbols))
```

Comment:
when the symbols are not actually present and they are retrieved asynchronously we should manually refresh the modeline to force building the breadcrumb with latest data.